### DISCUSSION TITLE
Get available webcams?

### DISCUSSION THREAD

**Discussion by gro-david** [j20001970/GDMP#35] [Q&A] [answered]
Is there any way to find the available webcams? If not how could this be implemented or worked around? 
(I am a complete newbie to c++, have no experience, but would be happy to develop it myself if someone could point me where to get started)
Cheers, Dave

**Comment by j20001970** [comment]
There is currently no methods available for camera enumeration in GDMP. If you want just workaround, on Linux you can check the presence of `/dev/video0` for webcam device at index 0.

If you would like to implement a method for that, it would be heavily platform-specific. We need to know to what extend each target platform could provide about camera device, then define the interface and data structure for them.

Using PipeWire should be a good start on Linux desktop, Windows could use DirectShow or newer Media Foundation, for Android we should be able to query camera information easier with updated CameraX. That being said I have few experience about them, and have no time to implement for now, but we can still trying to get it work without exposing the method to Godot before the feature is complete.

**Comment by gro-david** [comment]
GDMP uses opencv libraries, doesent it? Even though it doesent have a builtin way of finding all webcams, with it you can use try - catch statements and and just loop  through camera indexes until you find one that doesen throw an error. (https://stackoverflow.com/questions/34197825/c-opencv-2-4-11-list-all-cameras). I was thinking about implementing this, myself, but honestly i have no experience with cpp and have no idea where to even get started. If you could just help me with which file i would need to edit, i could figure out the rest myself (hopefully). 
Cheers

**Comment by j20001970** [comment]
OpenCV implementation of camera helper is located at `GDMP/io/camera_helper_cv.cc`, while `camera_helper.cc` and `camera_helper.h` is the interface exposed to Godot and will be implemented by multiple backends.

From what I see in OpenCV docs, neither `cv::VideoCapture` constructor nor `cv::VideoCapture::open` would throw exception, maybe the solution you linked is for older version OpenCV, but I guess you can still use `cv::VideoCapture::isOpened` when iterate over camera index.

Note that currently only desktop (Linux, Windows) uses OpenCV backend for camera helper, if you implement the enumeration method for `CameraHelperCV`, it will not be available for Android (`CameraHelperAndroid`) and iOS (`CameraHelperIOS`).

**Reply by gro-david** [reply]
Thanks for your reply, I am only planning to develop my app for Desktop, but I will maybe also try OS X. I will also try develop this feature as soon as possible. Should I then open a PR or is this not relevant for the majority of the users?

**Reply by j20001970** [reply, marked as answer]
I believe being able to enumerate available cameras would be a highly requested features, but considering how high-level OpenCV abstract the interaction with webcam devices, I doubt implementing webcam enumeration is suitable using OpenCV in terms of efficiency and reliability. My future plan is to implement different backends for different desktop platforms, while keeping CameraHelperCV as fallback. I wouldn't recommend to spend time creating PR for a backend that is considered fallback to-be.

**Reply by gro-david** [reply]
Okay, I will then only develop it for my own use, and publish it to my fork.
